### ISSUE TITLE
Backporting from effect/*

### ISSUE THREAD

**Issue by gcanti** [gcanti/fp-ts#1851]
- `flatMap` (aka chain / chainW) https://github.com/gcanti/fp-ts/pull/1849
- `tap` (aka chainFirst / chainFirstW)

**Comment by gcanti** [comment]
IMO it always makes sense since you get dual APIs and also a better name (tapError vs orElseFirst)

**Comment by sukovanej** [comment]
`as` and `asUnit` would be also nice, hm?

- `Either`
- `IO`
- `IOEither`
- `IOOption`
- `Option`
- `Reader`
- `ReaderEither`
- `ReaderIO`
- `ReaderTask`
- `ReaderTaskEither`

Does it make sense to implement `as` and `asUnit` for any data type having a functor instance or is there rule / law / practical reasoning for not having it for some data types? I see, in the `effect/data`, it's not implemented for arrays or `Identity`. What about `State` for example?

**Comment by gcanti** [comment]
No, there isn't a specific reason, they just don't seem very useful in the case of ReadonlyArray and Identity

**Comment by gcanti** [comment]
> Are struct and tuple from Product considered stable enough to be backported?

@sukovanej I think it's better to wait a little longer

p.s.
Anything else you would like to add? otherwise we could release 2.16

**Comment by sukovanej** [comment]
Ok. I don't have anything right now.

**Comment by christophgietl** [comment]
How about `tap` for `Identity`? For me it feels a little bit awkward to write `O.tap` and `I.chainFirst`.
